### PR TITLE
Feature/concurrent

### PR DESCRIPTION
**Pull request by memsharded**
I think it is mostly ready to be merged and released. Please review.

Can solve https://github.com/conan-io/conan/issues/1555

### CODE REVIEW

File: conans/client/client_cache.py [hunk 1]
```diff
@@ -36,6 +37,13 @@ def __init__(self, base_folder, store_folder, output):
         self._default_profile = None
         super(ClientCache, self).__init__(self._store_folder)
 
+    def conanfile_lock(self, conan_ref):
```
Comment (lasote) [inline conans/client/client_cache.py]: What folder exactly locks the `conanfile_lock`?

Comment (memsharded) [reply conans/client/client_cache.py]: The full package root.

Comment (lasote) [reply conans/client/client_cache.py]: A comment or rename containing "root" would help me.

File: conans/client/client_cache.py [hunk 2]
```diff
@@ -36,6 +39,16 @@ def __init__(self, base_folder, store_folder, output):
         self._default_profile = None
         super(ClientCache, self).__init__(self._store_folder)
 
+    def conanfile_read_lock(self, conan_ref):
```
Comment (lasote) [inline conans/client/client_cache.py]: Please put a comment: #lock for all the conanfile directory (root conanfile)

**Pull request merged**